### ISSUE TITLE
[BUG] Longhorn volume cannot attach to new node permanently if shutdown the current attached node. 

### ISSUE THREAD

**Issue by PhanLe1010** [longhorn/longhorn#9854]
## Describe the bug

This is a regression introduced in Longhorn `v1.6.3`

Share manager is permanently stuck in stopping/error if we shutdown the node of share manager pod. This leads to RWX PVC cannot attach to any new node.

Update: 
* This is also happen with RWO PVC

## To Reproduce

1. Deploy Longhorn `v1.6.3` 
1. Create a RWX PVC and a deployment of 1 pod using the PVC:
    <details close>
    <summary>Click to see the yaml</summary>

    <br>

    ```yaml
    apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      name: rwx-test
      namespace: default
    spec:
      accessModes:
        - ReadWriteMany
      storageClassName: longhorn
      resources:
        requests:
          storage: 1Gi
    ---
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: rwx-test
      labels:
        app: rwx-test
      namespace: default
    spec:
      replicas: 1
      selector:
        matchLabels:
          app: rwx-test
      strategy:
        type: Recreate
      template:
        metadata:
          labels:
            app: rwx-test
        spec:
          containers:
            - image: ubuntu:xenial
              imagePullPolicy: IfNotPresent
              command: [ "/bin/sh", "-c" ]
              args:
                - sleep 10; touch /data/index.html; while true; do date >> /data/index.html; sleep 1; done;
              name: rwx-test
              stdin: true
              tty: true
              volumeMounts:
                - mountPath: /data
                  name: rwx-test
            - image: nginx:stable
              imagePullPolicy: IfNotPresent
              name: nginx
              ports:
                - containerPort: 80
                  name: http
              volumeMounts:
                - mountPath: /usr/share/nginx/html
                  name: rwx-test
          restartPolicy: Always
          volumes:
            - name: rwx-test
              persistentVolumeClaim:
                claimName: rwx-test
    ```
    </details>
1. Find the node of the share manager pod inside longhorn-system namespace 
1. Shutdown that node
1. Scale up the deployment to 2 pods. Make sure new pod is running on a new node
1. Obverve that share manager CR stuck in stopping/error. Longhorn volume is stuck in detached or attached to the down node
1. Observe that the new pod cannot start because the PVC cannot be attached to the new node. Something similar to
    ```
    Events:
      Type     Reason                  Age                 From                     Message
      ----     ------                  ----                ----                     -------
      Normal   Scheduled               3m15s               default-scheduler        Successfully assigned default/rwx-test-65bdcb7b7-q2nbl to phan-v703-pool1-rdsql-kznvl
      Normal   SuccessfulAttachVolume  3m13s               attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-00579419-97a2-4341-84eb-12d156b18317"
      Warning  FailedMount             60s (x9 over 3m8s)  kubelet                  MountVolume.MountDevice failed for volume "pvc-00579419-97a2-4341-84eb-12d156b18317" : rpc error: code = InvalidArgument desc = volume pvc-00579419-97a2-4341-84eb-12d156b18317 hasn't been attached yet
    ```

## Expected behavior

RWX PVC should be able to attach to new node. Share manager CR should not stuck in stopping forever 

## RCA

1. When a node goes down, Longhorn relies on the Longhorn node's condition of type `ready` being set to `False` and the condition reason being set to `KubernetesNodeNotReady` to trigger the recover logic. Ref [code](https://github.com/longhorn/longhorn-manager/blob/442d68ac0046edf25a230ca414093397d89239ee/datastore/longhorn.go#L2724-L2743). A.K.A Longhorn expects the `nodes.longhorn.io` CR to be similar to this:
    ```
    status:
      autoEvicting: false
      conditions:
        - lastProbeTime: ''
          lastTransitionTime: '2024-11-26T00:53:15Z'
          message: >-
            Node phan-v703-pool1-rdsql-2x5bd is down: Kubernetes node XXX is not ready
          reason: KubernetesNodeNotReady
          status: 'False'
          type: Ready
    ```
1. The recovery flow includes:
    * Move ownership of the resources to new node
    * Force delete the share manager pod. Ref [code](https://github.com/longhorn/longhorn-manager/blob/442d68ac0046edf25a230ca414093397d89239ee/controller/share_manager_controller.go#L638-L645)
    * Instance Manager CR on the powered down node should be set to unknown. Ref [code](https://github.com/longhorn/longhorn-manager/blob/442d68ac0046edf25a230ca414093397d89239ee/controller/instance_manager_controller.go#L425-L440)
    * Engine and Replica on the down node are able to stopped
1. However, due to the change introduced in [this PR ](https://github.com/longhorn/longhorn-manager/pull/3112/files#diff-964ba26cdf441cbb013d7b13e1cfe460188feda96a4945b41d31009aef9a6931R1865-R1868) with this [github ticket](https://github.com/longhorn/longhorn/issues/7738), we never set the nodes condition reason being set to `KubernetesNodeNotReady`. The condition reason `ManagerPodDown` always shadow the condition reason `KubernetesNodeNotReady`

## Proposal

Do not make condition reason `ManagerPodDown` shadow the condition reason `KubernetesNodeNotReady`

**Comment by PhanLe1010** [comment]
It is quite difficult to workaround. We need the node to have KubernetesNodeNotReady condition to trigger the recover logic. But Longhorn manager are actively reset it back to ManagerPodDown

**Comment by PhanLe1010** [comment]
The issue also happens with RWO volume. A.K.A rendering all volumes on the down node unusable permanently. Volumes cannot move to new node unless the node come back.

**Comment by PhanLe1010** [comment]
### Test plan:

**Case 1:** (should be converted to an e2e test case)

1. Create a RWX PVC and a deployment of 1 pod using the PVC:
    <details close>
    <summary>Click to see the yaml</summary>

    <br>

    ```yaml
    apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      name: rwx-test
      namespace: default
    spec:
      accessModes:
        - ReadWriteMany
      storageClassName: longhorn
      resources:
        requests:
          storage: 1Gi
    ---
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: rwx-test
      labels:
        app: rwx-test
      namespace: default
    spec:
      replicas: 1
      selector:
        matchLabels:
          app: rwx-test
      strategy:
        type: Recreate
      template:
        metadata:
          labels:
            app: rwx-test
        spec:
          containers:
            - image: ubuntu:xenial
              imagePullPolicy: IfNotPresent
              command: [ "/bin/sh", "-c" ]
              args:
                - sleep 10; touch /data/index.html; while true; do date >> /data/index.html; sleep 1; done;
              name: rwx-test
              stdin: true
              tty: true
              volumeMounts:
                - mountPath: /data
                  name: rwx-test
            - image: nginx:stable
              imagePullPolicy: IfNotPresent
              name: nginx
              ports:
                - containerPort: 80
                  name: http
              volumeMounts:
                - mountPath: /usr/share/nginx/html
                  name: rwx-test
          restartPolicy: Always
          volumes:
            - name: rwx-test
              persistentVolumeClaim:
                claimName: rwx-test
    ```
    </details>
1. The Longhorn volume should have 3 replicas
1. Find the node of the share manager pod inside longhorn-system namespace 
1. Shutdown that node
1. Scale up the deployment to 2 pods. Make sure new pod is running on a new node

**Case 2:** (should be converted to an e2e test case)

Retest the RWX HA feature at https://github.com/james-munson/longhorn-tests/blob/2fc3a08cf2f20d7c79e786f65bd7b98aef206fbc/docs/content/manual/release-specific/v1.7.0/test-rwx-fast-failover.md

**Case 3:** (should be converted to an e2e test case)
1. Set setting `node-down-pod-deletion-policy` to `delete-both-statefulset-and-deployment-pod`
1. Create a RWO PVC and a deployment of 1 pod using the PVC
1. The volume should have 3 replicas
1. Shutdown that node which contains the pod
1. Verify that eventually a new replacement pod is running on a new node

**Comment by dfaltum** [comment]
I can't reproduce the problem on two different 1.7.1 test environments. The share-manager is recreated as expected.
Perhaps there are some Longhorn settings that prevent the problem?

**Comment by james-munson** [comment]
I think it is possible that fast-failover would not be affected by this issue for the reason that its decisions are based strictly on Lease status, not on node Conditions.  In fast failover, the share-manager typically has already been relocated to another node by the time that the restarted/powered off node even gets to NotReady.
There might be an exception in the case where the workload has only one replica (in the deployment sense) and it happens to be on the same node that was killed.

**Comment by PhanLe1010** [comment]
I am not very sure @james-munson because not only the removal of share manager pod is delayed but also the ownership transfer of the volume/engine/replica CRs.

Anyways, I think the action item is to make sure all test cases in this plan passes after the fix https://github.com/longhorn/longhorn/issues/9854#issuecomment-2499658139

**Comment by PhanLe1010** [comment]
> I can't reproduce the problem on two different 1.7.1 test environments. The share-manager is recreated as expected. Perhaps there are some Longhorn settings that prevent the problem?

You are right @dfaltum ! This issue shouldn't happen in v1.7.1 because v1.7.1 does not contains [this buggy commit](https://github.com/longhorn/longhorn-manager/pull/3138) which was merged only 1 day after v1.7.1 release (lucky us)

Sorry for the confusion, this regression should only exist in v1.6.3

**Comment by yangchiu** [comment]
Verification passed on master-head (longhorn-manager 2d5d202) following [the test plan](https://github.com/longhorn/longhorn/issues/9854#issuecomment-2499658139).
